gapis/replay/export_replay.go

```diff
@@ -86,43 +78,142 @@ func ExportReplay(ctx context.Context, pCapture *path.Capture, pDevice *path.Dev
+		state:   c.NewUninitializedState(ctx, ranges),
+		builder: b,
+	}
 
 	generatorReplayTimer.Time(func() {
-		err = generator.Replay(
+		ctx := status.Start(ctx, "Generate")
+		defer status.Finish(ctx)
+
+		err = m.key.generator.Replay(
 			ctx,
-			Intent{pDevice, pCapture},
-			Config(&struct{}{}),
-			[]RequestAndResult{{
-				Request: Request(generator.(interface{ ExportReplayRequest() Request }).ExportReplayRequest()),
-				Result:  func(val interface{}, err error) {},
-			}},
+			intent,
+			m.key.config,
+			m.requests,
 			d.Instance(),
 			c,
-			&adapter{
-				state:   c.NewUninitializedState(ctx, ranges),
-				builder: b,
-			})
+			out)
 	})
 
 	if err != nil {
-		return log.Err(ctx, err, "Replay returned error")
+		return nil, log.Err(ctx, err, "Replay returned error")
+	}
+
+	if config.DebugReplay {
+		log.I(ctx, "Building payload...")
 	}
 
 	var payload gapir.Payload
-	var handlePost builder.PostDataHandler
-	var handleNotification builder.NotificationHandler
-	builderBuildTimer.Time(func() { payload, handlePost, handleNotification, err = b.Build(ctx) })
+	builderBuildTimer.Time(func() { payload, _, _, err = b.Build(ctx) })
 	if err != nil {
-		return log.Err(ctx, err, "Failed to build replay payload")
+		return nil, log.Err(ctx, err, "Failed to build replay payload")
+	}
+	return &payload, nil
+}
+
+// NOTE: Only works for query that does not require result to be passed back.
+func (m *exportManager) Replay(
+	ctx context.Context,
+	intent Intent,
+	cfg Config,
+	req Request,
+	generator Generator,
+	hints *service.UsageHints) (val interface{}, err error) {
+
+	key := &batchKey{
+		capture:   intent.Capture.ID.ID(),
+		device:    intent.Device.ID.ID(),
+		config:    cfg,
+		generator: generator,
+	}
+
+	if m.key == nil {
+		m.key = key
+	}
+
+	if *key != *m.key {
+		return nil, log.Errf(ctx, nil, "Can not export trace with incompatible requests")
+	}
+
+	type res struct {
+		val interface{}
+		err error
+	}
+	out := make(chan res, 1)
+	m.requests = append(m.requests, RequestAndResult{
+		Request: req,
+		Result:  Result(func(val interface{}, err error) { out <- res{val, err} }),
+	})
+	return nil, nil
+}
+
+type ExportReplayFramebufferRequest struct {
+	Request          *service.GetFramebufferAttachmentRequest
+	Width, Height    uint32
+	FramebufferIndex uint32
+}
+
+// ExportReplay write replay commands and assets to path.
+func ExportReplay(ctx context.Context, pCapture *path.Capture, pDevice *path.Device, outDir string, fbs []ExportReplayFramebufferRequest) error {
+	if pDevice == nil {
+		return log.Errf(ctx, nil, "Unable to produce replay on unknown device.")
+	}
+
+	ctx = capture.Put(ctx, pCapture)
+	c, err := capture.Resolve(ctx)
+	if err != nil {
+		return err
+	}
+
+	mgr := &exportManager{}
+	for _, a := range c.APIs {
+		intent := Intent{pDevice, pCapture}
+		if a, ok := a.(QueryIssues); ok {
+			// If there are no frame query, force an issue query.
+			// Since otherwise the trace will not be generated.
```

Comment:
Is this a hack to generate a no-transform replay of the capture?
If so, please can we have a TODO to add a simple-replay mode here? GLES heavily transforms the capture for QueryIssues.

Reply:
Essentially yes, however I have not find a place in GAPID that a non-transformed trace is used.